### PR TITLE
[New Model] Product Stock

### LINKED ISSUE #145: [New Model]: ProductStock
## Model Description
The model should allow a supplier to provide a stock of his products (Bill to Order). The order number may not be known for sure and therefore should be optional.
Note: Issue#113 was to order specific. The information of stock for a partner should be oriented to materials.

## Relevant Standards
None

## Example Data
{
  "materialNumberCustomer": "MC-4711",
  "materialNumberSupplier": "MS-4711",
  "positions": [
    {
	  "positionId": "P-2023-03-04",
	  "lastUpdatedOn": "Iso Timestamp",
	  "customerOrderId": "OC-4711",
	  "supplierOrderId": "OS-4711",
	  "allocatedStock": {
	    "supplierStockLocation": {
		  "type": "BPNS", // BPNS, BPNA, Address
		  "value": "BPNS0123456789ZZ"
		  "quantityOnAllocatedStock": {
		    "quantityNumber": 2.5,
			"measurementUnit": "curie:litre"
		  }
		}
	  }
	}
  ]
}
## MS1 Criteria

- [ ] The proposed aspect model does not exist already
- [ ] The proposed aspect model does not extend an existing aspect but introduces completely new functionality
- [ ] The proposal references a Catena-X use case
- [ ] Relevant standards are mentioned/linked
- [ ] A first draft of the data structure is provided in an example

### PR DESCRIPTION
**Pull request by tom-rm-meyer-ISST**
## Description
This is the draft version for #145.

**Things to discuss:**
- materialNumberDataSpace: This shall be an optional reference to the PartAsPlanned. It should be aligend according to issue #103 - Therefore the data type is definetly wrong
- Location Entity: 
  - this should allow BPNS and BPNA (depending on how BPDM continues), should we add a trait to restrict the key type?
  - should we somewhat rename it to locationId?
  - should we set the see to the CX standard and if so, where?

Closes #145 

## MS2 Criteria
(to be filled out by PR reviewer)
- [x] the model **validates** with the BAMM SDS SDK in the version specified in the Readme.md of this repository by the time of the MS2 check  (e.g., 'java -jar bamm-cli.jar -i \<path-to-aspect-model\> -v ). The  BAMM CLI is available [here](https://openmanufacturingplatform.github.io/sds-documentation/sds-developer-guide/dev-snapshot/tooling-guide/bamm-cli.html) and in [GitHub](https://github.com/OpenManufacturingPlatform/sds-sdk/releases)
- [x] use **Camel-Case** (e.g., "MyModelElement" or "TimeDifferenceGmtId", when in doubt follow https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)
- [x] the identifiers for all model elements **start with a capital letter** except for properties
- [x] the identifier for **properties starts with a small letter**
- [x] all model elements **at least contain the fields "name" and "description"** in English language. 
- [x] **no duplicate names or preferredNames** within an Aspect (e.g. a Property and the referenced Characteristic should not have the same name)
- [x] the versioning in the URN **follows semantic versioning**, where minor version bumps are backwards compatible and major version bumps are not backwards compatible. 
- [x] use **abbreviations only when necessary** and if these are sufficiently common
- [x] **avoid redundant prefixes in property names** (consider adding properties to an enclosing Entity or even adapt the namespace of the model elements, e.g., instead of having two properties `DismantlerId` and `DismantlerName` use an Entity `Dismantler` with the properties `name` and `id` or use a URN like `io.catenax.dismantler:0.0.1`)
- [x] fields `preferredName` and `description` are not the same
- [x] **`preferredName` should be human readable** and follow normal orthography (e.g., no camel case but normal word separation)
- [x] name of aspect is singular except if it only has one property which is a Collection, List or Set. In theses cases, the aspect name is plural.
- [x] units are referenced from the BAMM unit catalog whenever possible
- [x] **use constraints** to make known constraints from the use case explicit in the aspect model 
- [x] when relying on **external standards**, they are referenced through a **"see"** element
- [x] all properties with an [simple type](https://openmanufacturingplatform.github.io/sds-documentation/bamm-specification/v1.0.0/datatypes.html) have an example value
- [x] metadata.json exists with status "release"
- [x] file RELEASE_NOTES.md exists and contains entries for proposed model changes 
- [x] all contributors to this model are mentioned in copyright header of model file

## MS3 Criteria
(to be filled out by semantic modeling team before merge to main-branch)
- [x] All required reviewers have approved this PR (see reviewers section)
- [x] The new aspect (version) will be implemented by at least one data provider
- [x] The new aspect (version) will be consumed by at least one data consumer
- [x] There exists valid test data
- [x] In case of a new (incompatible) major version to an existing version, a migration strategy has been developed
- [x] The model has at least version '1.0.0'

### CODE REVIEW

**Comment by tom-rm-meyer-ISST** [comment]
Modelling_Team

File: io.catenax.product_stock/1.0.0/metadata.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{ status : draft }
```
Comment (bs-jokri) [inline io.catenax.product_stock/1.0.0/metadata.json:1]: needs to be changed to release

**Review by bs-jokri — CHANGES_REQUESTED** [review]

File: io.catenax.product_stock/1.0.0/ProductStock.ttl [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    bamm:preferredName "Location Key Type Enum"@en;
+    bamm:description "Enumeration specifying types of keys for locations within Catena-X."@en;
+    bamm:dataType xsd:string;
+    bamm:see <urn:uri:CX-0010>;
+    bamm-c:values ("BPNS" "BPNA").
+:LocationKeyValueCharacteristic a bamm:Characteristic;
+    bamm:preferredName "Value for specific Location Key Characteristic"@en;
+    bamm:description "The actual key based on the type."@en;
+    bamm:dataType xsd:string;
+    bamm:see <urn:uri:CX-0010>.
```
Comment (bs-jokri) [inline io.catenax.product_stock/1.0.0/ProductStock.ttl:120]: same here. please check all the see elements and change to a proper uri (eg URL)

File: io.catenax.product_stock/1.0.0/ProductStock.ttl [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+:SupplierStockLocationIDCharacteristic a bamm:Characteristic;
+    bamm:preferredName "Location ID of the Supplier's Stock Characteristic"@en;
+    bamm:description "This is the location where the stock is located."@en;
+    bamm:dataType :LocationId.
+:LocationId a bamm:Entity;
+    bamm:properties (:locationIdType :locationId);
+    bamm:preferredName "Location ID"@en;
+    bamm:description "A location can be described by different kinds of identifiers. Within Catena-X, a location can either be described by a BPNS or BPNA."@en.
+:locationId a bamm:Property;
+    bamm:preferredName "ID value for specific Locatio"@en;
```
Comment (bs-jokri) [inline io.catenax.product_stock/1.0.0/ProductStock.ttl:135]: quite a long name. also fix Locatio to Location

File: io.catenax.product_stock/1.0.0/ProductStock.ttl [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    bamm:description "The actual ID reflecting the type (locationIdType) of the location."@en;
+    bamm:characteristic :LocationKeyValueCharacteristic;
+    bamm:exampleValue "BPNS0123456789ZZ".
+:locationIdType a bamm:Property;
+    bamm:preferredName "ype of specific Location ID"@en;
```
Comment (bs-jokri) [inline io.catenax.product_stock/1.0.0/ProductStock.ttl:140]: ype -> Type

File: io.catenax.product_stock/1.0.0/ProductStock.ttl [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    bamm:preferredName "Reference to Order Position"@en;
+    bamm:description "Encapsulates the references to identify a position within an order."@en.
+:DateTimeStamp a bamm:Characteristic;
+    bamm:preferredName "Date Time Stamp Characteristic"@en;
+    bamm:description "The Date with the the timezone of the involved site."@en;
+    bamm:dataType xsd:date;
+    bamm:see <urn:uri:ISO-8601>.
+:AllocatedStock a bamm:Entity;
+    bamm:properties (:quantityOnAllocatedStock :supplierStockLocationId);
+    bamm:preferredName "Stock allocated to Customer"@en;
```
Comment (bs-jokri) [inline io.catenax.product_stock/1.0.0/ProductStock.ttl:81]: check for consistent upper case writing in all preferredName tags. Sometimes you write lower case in the first character sometimes uppercase. I would opt for headline writing with uppercase writing of important words

**Review by bs-jokri — CHANGES_REQUESTED** [review]

**Comment by bs-jokri** [comment]
@tom-rm-meyer-ISST is there an update on the see topic?

**Comment by tom-rm-meyer-ISST** [comment]
> @tom-rm-meyer-ISST is there an update on the see topic?

No update by standardization team. Sent a friendly reminder.

From my point of view, we could resolve it here and update it with a patch later on.

**Comment by bs-jokri** [comment]
From MS3 review: materialNumberCustomer description needs to be adjusted to reflect that in some cases also the material number of the supplier can be used.

**Comment by petyog** [comment]
...
> Test data needs to be provided

A test data set can be provided upon request . Here is an example / snippet:
```json
{

  "materialNumberDataSpace" : "48878d48-6f1d-47f5-8ded-a441d0d879df",

  "materialNumberCustomer" : "MNR-7307-AU340474.002",

  "materialNumberSupplier" : "MNR-8101-ID146955.001",

  "positions" : [ {

    "lastUpdatedOnDate" : "2023-04-01T14:23:00.66372+01:00",

    "orderPositionReference" : {

      "supplierOrderId" : "M-Nbr-4711",

      "customerOrderId" : "C-Nbr-4711",

      "customerOrderPositionId" : "PositionId-01"

    },

    "allocatedStocks" : [ {

      "quantityOnAllocatedStock" : {

        "quantityNumber" : 2.5,

        "measurementUnit" : "unit:litre"

      },

      "supplierStockLocation" : {

        "locationKeyType" : "BPNS",

        "locationKeyValue" : "BPNS0123456789ZZ"

      }

    } ]

  } ]

}
```

**Comment by petyog** [comment]
@bs-jokri please don't merge the PR yet. There are still pending changes (the names of few fields must be streamlined). The team will get back to you once ready.

**Comment by petyog** [comment]
@bs-jokri turned out that there are no pending changes. The PR can be merged.

**Review by bs-jokri — APPROVED** [review]

**Pull request merged**